crates/hotshot/src/tasks/mod.rs

```diff
@@ -408,6 +410,91 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> EventTransformerState<TYPES,
     }
 }
 
+#[derive(Debug)]
+/// An `EventHandlerState` that modifies justify_qc to that of previous view to mock dishonest leader
+pub struct DishonestLeader<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    /// store events from previous views
+    pub validated_proposals: Vec<QuorumProposal<TYPES>>,
+    /// how many times current node has been elected leader and sent proposal
+    pub total_proposals_from_node: u64,
+    /// which proposals to be dishonest at
+    pub dishonest_at_proposal_numbers: HashSet<u64>,
+    /// how far back to look for a QC
+    pub view_look_back: usize,
+    /// phantom
+    pub _phantom: std::marker::PhantomData<I>,
+}
+
+/// add trait to handle proposal events for a dishonest leader
+trait DishonestProposal<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    /// send dishonest proposal
```

Comment:
Same here, this comment and function name somewhat clash. It looks like it does more than just send a single dishonest proposal.